### DISCUSSION TITLE
401 error when trying to run forge-pull

### DISCUSSION THREAD

**Discussion by LucieBlanvillain** [magit/forge#433] [Support] [answered]
Running forge-pull gives me a 401 "unauthorized", "/graphql" error with the message "bad credentials".
However, I'm sure my .authinfo.gpg file is set properly and readable by emacs, so I don't get where this error is coming from. Is there a way to check which credentials forge uses so I can make sure it's using the token I provided?

**Comment by LucieBlanvillain** [comment]
Update, I moved my configuration around a bit, eventually going back to the initial state, and now running forge-pull, pushing to / pulling from a repository etc. prompts for my username and password.

**Comment by tarsius** [comment]
> Is there a way to check which credentials forge uses so I can make sure it's using the token I provided?

Try `(ghub--token "api.github.com" "LucieBlanvill" 'forge)`.

However the Github web interface shows the token string only right after creating it, so there is no way for you to verify that this token is accepted except to try to make a request using it (and that indicates that it is *not* accepted). So I would recommend creating a new token (while making sure to use the correct permissions) and then pasting that into `.authinfo.gpg`. After you have done that you should restart Emacs because cached, and now incorrect, data is involved.

**Reply by LucieBlanvillain** [reply]
So this does show me the right token, however when I run forge commands I still get the prompt to login to github, so my username and token aren't being used.

**Reply by LucieBlanvillain** [reply]
Ok so as it turns out, I get a 401 error when my token is invalid, but when it is valid, it prompts me for my username and token.

**Reply by tarsius** [reply]
What is the prompt when you are asked for your username?

**Reply by LucieBlanvillain** [reply]
It says
"Username for 'https://github.com': "
then 
"Password for 'https://LucieBlanvillain@github.com': "

**Reply by c3d** [reply]
I'm curious what could cause the above to return a key that is different from the one stored in my `authinfo.gpg`?

**Reply by c3d** [reply, marked as answer]
> I'm curious what could cause the above to return a key that is different from the one stored in my authinfo.gpg?

Answering my own question: I apparently had a stray `.authinfo`. No idea where that came from.

**Comment by tarsius** [comment]
You are not supposed to ever get that prompt. See https://github.com/magit/forge/issues/293#issuecomment-728828720 for why. Ghub has a work around for that but apparently it does not work for you. Maybe you use a very old ghub version. Or maybe something else is wrong, I don't know.

So now we have three issues. The work around to prevent the next issue does not work. When something goes wrong, then you get username and password prompts instead of a meaningful error message because of an emacs bug. You have misconfigured something.

It it is very hard for me to diagnose what exactly you have done wrong from a distance. There's also a 5% that it isn't actually you who has done something wrong. I don't have the energy to help users one on one with such issues anymore. It is a huge emotional drain.

You should be able to get this working on your own by very carefully retracing each step, while consulting the documentation.

I now plan to write a debugging tool similar to the one I have written for `with-editor`, which would hopefully allow me to spot common mistakes without having to have a long conversation every time someone broke something.

I realize I am being a bad tempered here, but I just cannot take it anymore. Sorry for that. I hope you can understand.
